### PR TITLE
core: clean up warnings

### PR DESCRIPTION
**Pull request by BreadFish64**
cleans up all the errors encountered in core using MSVC ~~except a copious number of
`Warning	C4244	'initializing': conversion from 'const wchar_t' to 'char', possible loss of data	core	c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.14.26428\include\xmemory0	228	`~~

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3985)

### CODE REVIEW

File: src/core/arm/dyncom/arm_dyncom_trans.h [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#pragma warning(disable : 4200)
```
Comment (wwylele) [inline src/core/arm/dyncom/arm_dyncom_trans.h]: 1. comment what warning is this
2. wrap this in MSVC macro `#if`

File: src/core/core_timing.cpp [hunk 2]
```diff
@@ -122,7 +122,7 @@ u64 GetTicks() {
 }
 
 void AddTicks(u64 ticks) {
-    downcount -= ticks;
+    downcount -= static_cast<int>(ticks);
```
Comment (wwylele) [inline src/core/core_timing.cpp]: You should promote the type of `downcount` to `s64` instead

File: src/core/core_timing.cpp [hunk 3]
```diff
@@ -208,7 +208,7 @@ void Advance() {
         Event evt = std::move(event_queue.front());
         std::pop_heap(event_queue.begin(), event_queue.end(), std::greater<Event>());
         event_queue.pop_back();
-        evt.type->callback(evt.userdata, global_timer - evt.time);
+        evt.type->callback(evt.userdata, static_cast<int>(global_timer - evt.time));
```
Comment (wwylele) [inline src/core/core_timing.cpp]: Should cast to `s64`

File: src/core/file_sys/title_metadata.cpp [hunk 4]
```diff
@@ -73,7 +73,7 @@ Loader::ResultStatus TitleMetadata::Load(const std::vector<u8> file_data, size_t
     memcpy(&tmd_body, &file_data[offset + body_start], sizeof(TitleMetadata::Body));
 
     size_t expected_size =
-        body_start + sizeof(Body) + tmd_body.content_count * sizeof(ContentChunk);
+        body_start + sizeof(Body) + tmd_body.content_count * static_cast<u16>(sizeof(ContentChunk));
```
Comment (wwylele) [inline src/core/file_sys/title_metadata.cpp]: What does this do? isn't the result stored in `size_t`?

File: src/core/hle/kernel/ipc.cpp [hunk 5]
```diff
@@ -168,7 +169,7 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
             // page.
 
             if (!IsPageAligned(source_address) ||
-                (num_pages == 1 && !IsPageAligned(source_address + size))) {
+                (num_pages == 1 && !IsPageAligned(source_address + static_cast<VAddr>(size)))) {
```
Comment (wwylele) [inline src/core/hle/kernel/ipc.cpp]: Based on these three casting you should likely change the type of `size` to `u32` early.

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 6]
```diff
@@ -492,7 +492,7 @@ void HandleDeauthenticationFrame(const Network::WifiPacket& packet) {
         return;
     }
 
-    u16 node_id = node_map[packet.transmitter_address];
+    u16 node_id = static_cast<u16>(node_map[packet.transmitter_address]);
```
Comment (wwylele) [inline src/core/hle/service/nwm/nwm_uds.cpp]: Consider changing the type of `node_map` itself (cc @B3n30 )

File: src/core/hle/service/ptm/ptm.cpp [hunk 7]
```diff
@@ -77,7 +77,7 @@ void Module::Interface::GetStepHistory(Kernel::HLERequestContext& ctx) {
 
     u32 hours = rp.Pop<u32>();
     u64 start_time = rp.Pop<u64>();
-    size_t steps_buff_size;
+    u32 steps_buff_size = rp.Pop<u32>();
```
Comment (wwylele) [inline src/core/hle/service/ptm/ptm.cpp]: Where does this come from?

File: src/core/hle/service/soc_u.cpp [hunk 8]
```diff
@@ -796,8 +796,8 @@ void SOC_U::SetSockOpt(Kernel::HLERequestContext& ctx) {
 #endif
     } else {
         const char* optval_data = reinterpret_cast<const char*>(optval.data());
-        err = static_cast<u32>(
-            ::setsockopt(socket_handle, level, optname, optval_data, optval.size()));
+        err = static_cast<u32>(::setsockopt(socket_handle, level, optname, optval_data,
+                                            static_cast<int>(optval.size())));
```
Comment (wwylele) [inline src/core/hle/service/soc_u.cpp]: The type of this on linux is `socklen_t` (unsigned at least u32)

File: src/core/hle/service/ptm/ptm.cpp (hunk 7 above)
Comment (BreadFish64) [reply src/core/hle/service/ptm/ptm.cpp]: It was an unreferenced variable, in the header it says it's a buffer descriptor and in APT it's popped this way

Comment (wwylele) [reply src/core/hle/service/ptm/ptm.cpp]: since it is a buffer descriptor it is already propped along with the mapped buffer below. Just remove this.

Comment (wwylele) [reply src/core/hle/service/ptm/ptm.cpp]: Could you reference me the APT code?

Comment (BreadFish64) [reply src/core/hle/service/ptm/ptm.cpp]: nvm

File: src/core/file_sys/title_metadata.cpp [hunk 9]
```diff
@@ -72,8 +72,8 @@ Loader::ResultStatus TitleMetadata::Load(const std::vector<u8> file_data, size_t
     memcpy(tmd_signature.data(), &file_data[offset + sizeof(u32_be)], signature_size);
     memcpy(&tmd_body, &file_data[offset + body_start], sizeof(TitleMetadata::Body));
 
-    size_t expected_size =
-        body_start + sizeof(Body) + tmd_body.content_count * sizeof(ContentChunk);
+    size_t expected_size = body_start + sizeof(Body) +
+                           static_cast<size_t>(tmd_body.content_count) * sizeof(ContentChunk);
```
Comment (wwylele) [inline src/core/file_sys/title_metadata.cpp]: `tmd_body.content_count` should be converted to `u16` first (BE->native conversion) and then converted to `size_t`. The second conversion can be implicit as the `*` operator should be able to promote it.

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 10]
```diff
@@ -612,7 +612,7 @@ void NWM_UDS::RecvBeaconBroadcastData(Kernel::HLERequestContext& ctx) {
     }
 
     // Update the total size in the structure and write it to the buffer again.
-    data_reply_header.total_size = cur_buffer_size;
+    data_reply_header.total_size = static_cast<u32_le>(cur_buffer_size);
```
Comment (wwylele) [inline src/core/hle/service/nwm/nwm_uds.cpp]: only cast to `u32` and let the swap type do implicit cast (if necessary)

**Comment by BreadFish64** [comment]
updating fmt to the latest stable release has fixed the aforementioned C4244 warnings

**Comment by BreadFish64** [comment]
rebased

**Pull request merged**